### PR TITLE
STL Viewer and Combiner 

### PR DESCRIPTION
**Pull request by ArchanGhosh**
# Description

A small script that allows us to view a STL file in console and also combine Multiple STL file without the need of an actual 3d application like Blender or Maya

## Type of change

_Please delete options that are not relevant._

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation Update

# Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests/screenshots(if any) that prove my fix is effective or that my feature works

### CODE REVIEW

File: Python/View_and_Combine_STL/view.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+	figure = pyplot.figure(figsize=(30,30))
+	axes = mplot3d.Axes3D(figure)
+
+	view_mesh = mesh.Mesh.from_file(file_dir)
+	axes.add_collection3d(mplot3d.art3d.Poly3DCollection(your_mesh.vectors))
+
+	scale = view_mesh.points.flatten()
+	axes.auto_scale_xyz(scale, scale, scale)
+
+	pyplot.show()
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/view.py:15]: Please add a line at the EOF.

File: Python/View_and_Combine_STL/view.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from stl import mesh
+from mpl_toolkits import mplot3d
+from matplotlib import pyplot
+
+def view_stl(file_dir):
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/view.py:5]: Add a function docstring to document what the function is doing

File: Python/View_and_Combine_STL/view.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+def view_stl(file_dir):
+	figure = pyplot.figure(figsize=(30,30))
+	axes = mplot3d.Axes3D(figure)
+
+	view_mesh = mesh.Mesh.from_file(file_dir)
+	axes.add_collection3d(mplot3d.art3d.Poly3DCollection(your_mesh.vectors))
+
+	scale = view_mesh.points.flatten()
+	axes.auto_scale_xyz(scale, scale, scale)
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/view.py:13]: Can you add comments here to show what exactly they are doing?

File: Python/View_and_Combine_STL/combine.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import math
+import stl
+from stl import mesh
+import numpy
+import glob
+
+def combine_stl(data_dir):
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/combine.py:7]: Add a function docstring

File: Python/View_and_Combine_STL/README.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# View and Combine STL
+
+STL files are an important part of 3D asset creation and management. STL files provide us with basic maps of 3D objects created/edited in applications like Blender/Unreal/Unity
+
+STL are mainly used to create 3D maps of environments like cityscapes and heightmaps of terrain. 
+
+## Running the Script
+* Create a virtual env
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/README.md:8]: Replace `Create a virtual env` to `Initialize and start a Virtual Environment`. If possible you can add a good resource on the topic as well as a hyperlink

File: Python/View_and_Combine_STL/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+* install numpy-stl, numpy and matplotlib or use requirements.txt
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/README.md:9]: Refer to the proper command: `Install the Packages by running the command: pip install -r requirements.txt`

File: Python/View_and_Combine_STL/README.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+* You can use combine.py to concatenate multiple STL files(recommended for assets designed in same 3D space or domain)
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/README.md:10]: Please explain how.  If possible add a sample I/O as a Code Block

File: Python/View_and_Combine_STL/README.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+* Or you can simply run view.py to view the STL file in a pyplot env.(set the target size to your liking)
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/README.md:11]: Please explain why. You can consider that I'm a beginner with no idea whatsoever, and want to use this. How am I going to use it?

File: Python/View_and_Combine_STL/README.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+When you use City Generator, kindly delete or store the domain.stl file before running the combine.py script as it will result in the creation of flat mesh instead of a 3D one.
+
+Further you can use [Blender](https://www.blender.org/) which is one of the most amazing and Open Source 3D asset and animation creation tool out there.
+
+![3D Cityscape in paint3D](https://github.com/HarshCasper/Rotten-Scripts/blob/master/Python/View_and_Combine_STL/images/building_3d.png)
+![STL View in Python](https://github.com/HarshCasper/Rotten-Scripts/blob/master/Python/View_and_Combine_STL/images/building_py_view.png)
+
+![Road 3D](https://github.com/HarshCasper/Rotten-Scripts/blob/master/Python/View_and_Combine_STL/images/road_3d.PNG)
+![Road Py View](https://github.com/HarshCasper/Rotten-Scripts/blob/master/Python/View_and_Combine_STL/images/road_py_view.PNG)
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/README.md:24]: Please upload all your images on Imgur and reference it here

File: Python/README.md [hunk 10]
```diff
@@ -337,20 +338,21 @@ S. No | Name of Script | Directory Link |  Author | Tools Used |
...
+159 | Word Meaning | [Click Here](/Python/Word_Meaning) | [Mohit Gupta](https://github.com/mohitacecode) | PyDictionary
+160 | Wifi Spped Tester | [Click Here](/Python/Wifi%20Speed%20Tester) | [Vedansh Vijaywargiya](https://github.com/vvHacker007) | win10toast, speedtest
+161 | XML To CSV | [Click Here](/Python/XML_To_CSV) | [Harsh Bardhan Mishra](https://github.com/HarshCasper) | glob, pandas
+162 | Youtube Bot | [Click Here](/Python/Youtube_Bot) | [Sri Manikanta Palakollu](https://github.com/srimani-programmer) | time, argparse, selenium
+163 | Youtube Video Download | [Click Here](/Python/Youtube_Video_Download) | [Anushka-S10](https://github.com/Anushka-S10) | pytube
+164 | YouTube Video-Audio Download | [Click Here](/Python/YouTube_Video-Audio_Download) | [Naman Shah](https://github.com/namanshah01) | os, sys, pafy, requests
+165 | Zipper | [Click Here](/Python/Zipper) | [Puneet Pal Singh](https://github.com/Puneet-Pal-Singh) | os, zipfile, datetime
+166 | Zoom Automation | [Click Here](/Python/Zoom_Automation) | [Rutuja Dhanawade](https://github.com/rutujadhanawade) | pyautogui, time
```
Comment (HarshCasper) [inline Python/README.md:357]: You can revert your commits for the Python README. I have delivered a fix right now, which might create some merge conflicts. Its better to give it a skip and come up with a more friendly approach of sorting them out.

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Thanks for the exceptional work! 

Few changes requested.

File: Python/README.md (hunk 10 above)
Comment (ArchanGhosh) [reply Python/README.md]: Changes Reverted

File: Python/View_and_Combine_STL/README.md (hunk 5 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/README.md:8]: Resolved

File: Python/View_and_Combine_STL/README.md (hunk 6 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/README.md:9]: Resolved

File: Python/View_and_Combine_STL/README.md (hunk 9 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/README.md:24]: Actually I tried using Imgur before this but for some reason the embedding didn't work and the images were not loading.

File: Python/View_and_Combine_STL/view.py (hunk 3 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/view.py:13]: Resolved

File: Python/View_and_Combine_STL/view.py (hunk 2 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/view.py:5]: Resolved

File: Python/View_and_Combine_STL/combine.py (hunk 4 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/combine.py:7]: Resolved

File: Python/View_and_Combine_STL/view.py (hunk 1 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/view.py]: Resolved

File: Python/View_and_Combine_STL/README.md (hunk 7 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/README.md]: Resolved

File: Python/View_and_Combine_STL/README.md (hunk 9 above)
Comment (HarshCasper) [reply Python/View_and_Combine_STL/README.md:49]: Can you please have a look at other PRs that have done the same thing. 

Please go to this [link](https://imgur.com/upload), upload the Image and take the Image URL and add it above.

File: Python/View_and_Combine_STL/combine.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import numpy
+import glob
+
+def combine_stl():
+'''This function combines all the STL file in a directory and merges them together
+   Params: data_dir = stores the directory of stl files
+   	   stl_dir = list all the stl files
+	   data = total number of vertices and their value
+	   combine = Combines all the stl files relative to their location as placed originally 
+'''
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/combine.py:13]: Not a very proper way of adding Docstrings

File: Python/View_and_Combine_STL/view.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+from matplotlib import pyplot
+
+def view_stl():
+"""The following function renders a 3D geometry mesh in Matplotlib, if helps you visualize a 3D object in a simple and fast way
+   Params: file_dir = to store the directory location
+   	   figure = to create an empty Pyplot figure
+	   axes = subgraph to plot the mesh
+	   view_mesh = stores the 3D mesh from the directory
+	   scale = stores the relative scale of vertices:plot size
+"""
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/view.py:12]: Not a very proper way of adding docstrings

File: Python/View_and_Combine_STL/combine.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+	#Creating an Empty mesh to concatenate all the stl file in a directory
+	data = numpy.zeros(0, dtype=mesh.Mesh.dtype)
+	combine = mesh.Mesh(data, remove_empty_areas=False)
+	
+	files = glob.glob(stl_dir)
+
+	for fl in files:
+  		stl_fl = mesh.Mesh.from_file(fl)
+  		combine = mesh.Mesh(numpy.concatenate([stl_fl.data, combine.data]))
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/combine.py:26]: The indentation looks inconsistent

File: Python/View_and_Combine_STL/view.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+	file_dir = "REPLACE WITH THE LOCATION OF THE FILE" #example "c:\users\username\..."
+	figure = pyplot.figure(figsize=(30,30))
+	axes = mplot3d.Axes3D(figure)
+
+	view_mesh = mesh.Mesh.from_file(file_dir)
+	axes.add_collection3d(mplot3d.art3d.Poly3DCollection(your_mesh.vectors))
+	
+	#Here we calculated the total number of flat vertices, that are present in a 3D object and then scale it according to the total Figsize of the plot
+	#Autoscale allows us to fit the object within the constraints of a PyPlot Graph, and also fixes the object at 0,0,0 coordinates which 
+	#can help us understand the true dimensions of the object in inches
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/view.py:22]: Keep it in a multi-line strong and remove trailing whitespace

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: Python/View_and_Combine_STL/README.md (hunk 9 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/README.md]: Resolved

File: Python/View_and_Combine_STL/README.md [hunk 15]
```diff
@@ -0,0 +1,49 @@
+# View and Combine STL
+
+STL files are an important part of 3D asset creation and management. STL files provide us with basic maps of 3D objects created/edited in applications like Blender/Maya/Unreal/Unity
+
+STL are mainly used to create 3D maps of environments like cityscapes and heightmaps of terrain. 
+
+## Running the Script
+* Initialize and start a Virtual Environment
+* Install the Packages by running the command: pip install -r requirements.txt
+* Running Combine.py will combine all the STL files present in the drectory referred to.
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/README.md:10]: Keep <b>Combine.p</b> in backticks

File: Python/View_and_Combine_STL/README.md [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+* Running View.py will let you view the any STL file in a 3D PyPlot Space.
```
Comment (HarshCasper) [inline Python/View_and_Combine_STL/README.md:11]: Keep <b>View.py</b> in backticks

File: Python/View_and_Combine_STL/view.py (hunk 14 above)
Comment (HarshCasper) [reply Python/View_and_Combine_STL/view.py:22]: Please resolve this comment as well

File: Python/View_and_Combine_STL/combine.py (hunk 13 above)
Comment (HarshCasper) [reply Python/View_and_Combine_STL/combine.py:26]: Can you have a look at the same?

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Remove all the Images as well from the directory.

File: Python/README.md [hunk 17]
```diff
@@ -172,4 +172,4 @@ S. No | Name of Script | Directory Link |  Author | Tools Used |
 162 | Youtube Video Download | [Click Here](/Python/Youtube_Video_Download) | [Anushka-S10](https://github.com/Anushka-S10) | pytube
 163 | YouTube Video-Audio Download | [Click Here](/Python/YouTube_Video-Audio_Download) | [Naman Shah](https://github.com/namanshah01) | os, sys, pafy, requests
 164 | Zipper | [Click Here](/Python/Zipper) | [Puneet Pal Singh](https://github.com/Puneet-Pal-Singh) | os, zipfile, datetime
-165 | Zoom Automation | [Click Here](/Python/Zoom_Automation) | [Rutuja Dhanawade](https://github.com/rutujadhanawade) | pyautogui, time
+165 | Zoom Automation | [Click Here](/Python/Zoom_Automation) | [Rutuja Dhanawade](https://github.com/rutujadhanawade) | pyautogui, time
```
[the remark this answers is not in the document]
Comment (ArchanGhosh) [reply Python/README.md:175]: Resolved

**Comment by ArchanGhosh** [comment]
> Remove all the Images as well from the directory.

Removed

File: Python/View_and_Combine_STL/README.md (hunk 15 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/README.md]: Resolved

File: Python/View_and_Combine_STL/combine.py (hunk 13 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/combine.py:26]: This is the default tab space/indentation for my editior

File: Python/View_and_Combine_STL/README.md [hunk 18]
```diff
@@ -0,0 +1,49 @@
+# View and Combine STL
+
+STL files are an important part of 3D asset creation and management. STL files provide us with basic maps of 3D objects created/edited in applications like Blender/Maya/Unreal/Unity
+
+STL are mainly used to create 3D maps of environments like cityscapes and heightmaps of terrain. 
+
+## Running the Script
+* Initialize and start a Virtual Environment
```
[the remark this answers is not in the document]
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/README.md:8]: Resolved

File: Python/View_and_Combine_STL/view.py (hunk 14 above)
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/view.py]: Resolved

File: Python/View_and_Combine_STL/README.md [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+* Install the Packages by running the command: pip install -r requirements.txt
```
[the remark this answers is not in the document]
Comment (ArchanGhosh) [reply Python/View_and_Combine_STL/README.md]: Resolved

File: Python/README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (HarshCasper) [reply Python/README.md:175]: It has not been resolved.

File: Python/View_and_Combine_STL/README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (HarshCasper) [reply Python/View_and_Combine_STL/README.md:8]: It has not been resolved.

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Tiny changes requested. Once done, squash all your commits into one single commit and push it back. It will be merged.

**Comment by vybhav72954** [comment]
@ArchanGhosh This is one wonderful and interesting script. Kindly squash the commits so that we can get this merged.

**Comment by vybhav72954** [comment]
@HarshCasper Can we squash the commits while merging from our side only?

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**